fineract-progressive-loan/src/main/java/org/apache/fineract/portfolio/loanaccount/service/ProgressiveLoanInterestRefundServiceImpl.java

```diff
@@ -0,0 +1,130 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements. See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership. The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.fineract.portfolio.loanaccount.service;
+
+import java.math.BigDecimal;
+import java.time.LocalDate;
+import java.util.List;
+import java.util.concurrent.atomic.AtomicReference;
+import lombok.RequiredArgsConstructor;
+import lombok.extern.slf4j.Slf4j;
+import org.apache.fineract.organisation.monetary.domain.MonetaryCurrency;
+import org.apache.fineract.portfolio.loanaccount.domain.Loan;
+import org.apache.fineract.portfolio.loanaccount.domain.LoanRepaymentScheduleInstallment;
+import org.apache.fineract.portfolio.loanaccount.domain.LoanTransaction;
+import org.apache.fineract.portfolio.loanaccount.domain.transactionprocessor.impl.AdvancedPaymentScheduleTransactionProcessor;
+import org.apache.fineract.portfolio.loanaccount.loanschedule.data.ProgressiveLoanInterestScheduleModel;
+import org.apache.fineract.portfolio.loanaccount.starter.AdvancedPaymentScheduleTransactionProcessorCondition;
+import org.apache.fineract.portfolio.loanproduct.calc.EMICalculator;
+import org.springframework.context.annotation.Conditional;
+import org.springframework.stereotype.Service;
+
+@Slf4j
+@RequiredArgsConstructor
+@Conditional(AdvancedPaymentScheduleTransactionProcessorCondition.class)
+@Service
+public class ProgressiveLoanInterestRefundServiceImpl implements InterestRefundService {
+
+    private final AdvancedPaymentScheduleTransactionProcessor processor;
+    private final EMICalculator emiCalculator;
+
+    private static LoanRepaymentScheduleInstallment duplicate(LoanRepaymentScheduleInstallment rs) {
+        MonetaryCurrency currency = rs.getLoan().getCurrency();
+
+        return new LoanRepaymentScheduleInstallment(rs.getLoan(), rs.getInstallmentNumber(), rs.getFromDate(), rs.getDueDate(),
+                rs.getPrincipal(currency).getAmount(), rs.getInterestCharged(currency).getAmount(),
+                rs.getFeeChargesCharged(currency).getAmount(), rs.getPenaltyChargesCharged(currency).getAmount(),
+                rs.isRecalculatedInterestComponent(), rs.getLoanCompoundingDetails(), rs.getRescheduleInterestPortion(),
+                rs.isDownPayment());
+    }
```

Comment:
I'd say this would be rather a responsibility of a separate class, what do you think?